crates/task/src/task.rs

```diff
@@ -1,81 +1,45 @@
 use async_std::{
     sync::RwLock,
     task::{spawn, JoinHandle},
 };
+use async_trait::async_trait;
 #[cfg(async_executor_impl = "async-std")]
 use futures::future::join_all;
 #[cfg(async_executor_impl = "tokio")]
 use futures::future::try_join_all;
-use futures::{future::select_all, Future};
 #[cfg(async_executor_impl = "tokio")]
 use tokio::{
     sync::RwLock,
     task::{spawn, JoinHandle},
 };
-use tracing::{error, warn};
 
-use crate::{
-    dependency::Dependency,
-    dependency_task::{DependencyTask, HandleDepOutput},
-};
+/// Trait for events that long-running tasks handle
+pub trait TaskEvent: PartialEq {
+    /// The shutdown signal for this event type
+    fn shutdown_event() -> Self;
+}
 
+#[async_trait]
 /// Type for mutable task state that can be used as the state for a `Task`
 pub trait TaskState: Send {
     /// Type of event sent and received by the task
-    type Event: Clone + Send + Sync + 'static;
-    /// The result returned when this task completes
-    type Output: Send;
-    /// Handle event and update state.  Return true if the task is finished
-    /// false otherwise.  The handler can access the state through `Task::state_mut`
-    fn handle_event(
-        event: Self::Event,
-        task: &mut Task<Self>,
-    ) -> impl Future<Output = Option<Self::Output>> + Send
-    where
-        Self: Sized;
+    type Event: TaskEvent + Clone + Send + Sync;
 
-    /// Return true if the event should be filtered
-    fn filter(&self, _event: &Self::Event) -> bool {
-        // default doesn't filter
-        false
-    }
-    /// Do something with the result of the task before it shuts down
-    fn handle_result(&self, _res: &Self::Output) -> impl std::future::Future<Output = ()> + Send {
-        async {}
-    }
-    /// Return true if the event should shut the task down
-    fn should_shutdown(event: &Self::Event) -> bool;
-    /// Handle anything before the task is completely shutdown
-    fn shutdown(&mut self) -> impl std::future::Future<Output = ()> + Send {
-        async {}
-    }
-}
+    /// Joins all subtasks.
+    async fn cancel_subtasks(&mut self);
 
-/// Task state for a test.  Similar to `TaskState` but it handles
-/// messages as well as events.  Messages are events that are
-/// external to this task.  (i.e. a test message would be an event from non test task)
-/// This is used as state for `TestTask` and messages can come from many
-/// different input streams.
-pub trait TestTaskState: Send {
-    /// Message type handled by the task
-    type Message: Clone + Send + Sync + 'static;
-    /// Result returned by the test task on completion
-    type Output: Send;
-    /// The state type
-    type State: TaskState;
-    /// Handle and incoming message and return `Some` if the task is finished
-    fn handle_message(
-        message: Self::Message,
-        id: usize,
-        task: &mut TestTask<Self::State, Self>,
-    ) -> impl Future<Output = Option<Self::Output>> + Send
-    where
-        Self: Sized;
```

Comment:
`TestTaskState` and `TestTask` are moved to `testing/src/test_task.rs`.

The trait and struct themselves have been somewhat reworked, but should fit into the test runner in the same way.